pkg/apis/flowcontrol/types.go

```diff
@@ -517,3 +530,14 @@ const (
 	ConditionFalse   ConditionStatus = "False"
 	ConditionUnknown ConditionStatus = "Unknown"
 )
+
+// ConcurrencyLimitStatus reports the limit imposed by one apiserver.
+type ConcurrencyLimitStatus struct {
```

Comment:
You need a way to remove these when they're no longer being updated.  https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2339-storageversion-api-for-ha-api-servers#garbage-collection is a similar case.

Reply:
impl leaves something to be desired: https://github.com/kubernetes/kubernetes/blob/d13fe474f9c9e42ac88e22985c503ab7217329db/pkg/controller/storageversiongc/gc_controller.go#L261

Reply:
Actually I have a deeper question.  There is already another controller detecting when kube-apiservers go way, it is part of the lease-based technique for maintaining the kube-apiservers Endpoints.  The controller doing GC of storageversions is riding on the lease-based Endoints behavior, but indirectly.  And yeah, it's bogus to put an object rather than an object reference in the downstream controller's workqueue.  But the deeper question is: why be so indirect?  It would be wrong to add another copy of this downstream detection code.  Wouldn't it be better to enhance the originating controller so that it can be directly connected to clients?  For example, suppose the lease-based kube-apiserver Endpoints controller can be given some pipelines to clients.  A workqueue seems like an obvious choice for a pipeline to a client.  Let the lease-based kube-apiserver Endpoints controller send a [comparable](https://golang.org/ref/spec#Comparison_operators) representation of the set of existing kube-apiserver IDs upon introduction and every change.  The StorageVersion GC controller could be simplified by consuming that, and the PriorityLevelConfigurationStatus GC could also be a client of that.

Reply:
Actually a workqueue is not the right choice, its duplicate suppression does not preserve lastness.

Reply:
However, the current controller, which is in `pkg/controlplane/controller/apiserverleasegc/gc_controller.go`, does not attempt to promptly react to all positive edges (lease creations) --- it only needs to eventually be consistent with negative edges (lease expirations).

Reply:
The originating controller uses the public IP to identify individual masters, which doesn't work in all flavor of Kubernetes deployment, e.g. GKE masters are sitting behind a load balancer. 

> the current controller, which is in `pkg/controlplane/controller/apiserverleasegc/gc_controller.go`, does not attempt to promptly react to all positive edges (lease creations) --- it only needs to eventually be consistent with negative edges (lease expirations).

Is this behavior of current / originating controller not satisfying the requirement of PriorityLevelConfiguration? I think the current controller's behavior can be tuned to match the originating controller's. The originating controller periodically write new leases and GC expired leases. The current controller broke it into two parts: [gc_controller](https://github.com/kubernetes/kubernetes/blob/master/pkg/controlplane/controller/apiserverleasegc/gc_controller.go) and [lease_controller](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-helpers/apimachinery/lease/controller.go). GC doesn't have to react to lease creations IMO.

Reply:
@roycaihw : I am a little unsure of the tense of your remark that

> The originating controller uses the public IP to identify individual masters, which doesn't work in all flavor of Kubernetes deployment, e.g. GKE masters are sitting behind a load balancer. So we have to introduce a new unique ID

I am looking at https://github.com/kubernetes/kubernetes/blob/20b0f6b2a6ab9b1f0df821044150539ffc2c5503/pkg/controlplane/instance.go#L491-L522 and I do not see use of IP addresses.  I see identification by `m.GenericAPIServer.APIServerID`, which comes from a command line argument that has a default that is a random UUID plus a fixed prefix.  Perhaps you meant to explain the motivation for the code I see?

For removing deceased apiservers from PriorityLevelConfigurationStatus --- in a state-based way, which is how we prefer to write controllers --- the downstream controller (https://github.com/kubernetes/apiserver/blob/master/pkg/util/flowcontrol/apf_controller.go) needs to know all the live apiservers.  BTW, the existing StorageVersionGC controller is edge-based rather than state-based.  If it were state-based then it would need the same thing --- awareness of all the live apiservers.

My concern here is that it is wrong to have multiple copies of the same code.  There should be one place where the set of live apiservers is tracked, and the downstream controllers that need to know this set should be connected as clients rather than copies.

Reply:
> I do not see use of IP addresses

ah I'm talking about something different. Sorry about the confusion. I read 

> There is already another controller detecting when kube-apiservers go way, it is part of the lease-based technique for maintaining the kube-apiservers Endpoints

and I thought the "originating controller" was https://github.com/kubernetes/kubernetes/blob/20b0f6b2a6ab9b1f0df821044150539ffc2c5503/pkg/controlplane/reconcilers/lease.go#L151